modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -2409,18 +2394,23 @@ def get_axis_lengths(partitions, axis):
         do_reindex_base = not base_index.equals(joined_index)
         do_repartition_base = force_repartition or do_reindex_base
 
-        # perform repartitioning and reindexing for `base_frame` if needed
+        # perform repartitioning and reindexing for `base_frame` if needed.
+        # also define length of base and frames. we will need to know the
+        # lengths for alignment.
         if do_repartition_base:
             reindexed_base = base_frame._partition_mgr_cls.map_axis_partitions(
                 axis,
                 base_frame._partitions,
                 make_reindexer(do_reindex_base, base_frame_idx),
             )
+            if axis:
+                base_lengths = [obj.width() for obj in reindexed_base[0]]
```

Comment:
Can we use the `_column_widths` or `_row_lengths` attributes to compute these?

Reply:
`reindexed_base` is a 2-d array of frames, not a Modin dataframe, so it doesn't have those attributes. I instead made `_copartition` return the computed axis lengths. I've made all methods that call `_copartition` use the computed axis lengths when possible.